File: Talks/find-talks.py
#!/usr/bin/env python3
#
# A small collection script for talk description files (YAML).
# Heavily inspired by https://github.com/svenk/uniordner/tree/master/unibib-searchengine
#

import pathlib, yaml, sys
here = pathlib.Path(__file__).parent
here_pub = "https://github.com/svenk/publications/tree/master/Talks"
talks = []

def fixlinks(talk, path):
    "Discover anything file-like in the talk dict and replace it with public URL."
    base = path.parent
    base_pub = str(base).replace(str(here),here_pub)
    talk_pub = {}
    fixer = lambda v: base_pub+"/"+v if (base/str(v)).exists() else v
    for k,v in talk.items():
        if isinstance(v, dict):
            talk_pub[k] = fixlinks(v, path)
        elif isinstance(v, list):
            talk_pub[k] = [ fixer(w) for w in v ]
        else:
            talk_pub[k] = fixer(v)
    return talk_pub

for path in here.rglob("*.talk.y*ml"):
    #print(path, file=sys.stderr)
    content = yaml.safe_load(open(path, "r"))
    if isinstance(content, list):
        talks += (fixlinks(talk,path) for talk in content)
    elif isinstance(content, dict):
        for name, talk in content.items():
            talk['key'] = name
            talks.append(fixlinks(talk,path))

print(yaml.dump({"talks": talks }))

File: Talks/find-talks.js
#!/usr/bin/env node
// small standalone, needs npm install yaml and glob.
// This is basically a nodejs-variant of find-talks.py

import fs from "fs";
import path from "path";
import yaml from "js-yaml";
import { globSync } from "glob";
import { fileURLToPath } from "url";

const here = path.dirname(fileURLToPath(import.meta.url));
const here_pub = "https://github.com/svenk/publications/tree/master/Talks";
const talks = [];

const githubTreeDownload = v => {
  if (typeof v !== "string") return null;
  const m = v.match(/^https:\/\/github\.com\/([^/]+)\/([^/]+)\/tree\/([^/]+)\/(.+\.(?:pdf|odp|pptx?|key|zip))(?:[?#].*)?$/i);
  return m && `https://raw.githubusercontent.com/${m[1]}/${m[2]}/${m[3]}/${m[4]}`;
};

const fixlinks = (t, p) => {
  const b = path.dirname(p);
  const bp = b.replace(here, here_pub);
  const fx = v =>
    fs.existsSync(path.join(b, String(v))) ? bp + "/" + v : v;

  const o = {};
  for (const [k, v] of Object.entries(t)) {
    if (Array.isArray(v)) o[k] = v.map(fx);
    else if (v && typeof v === "object") o[k] = fixlinks(v, p);
    //else o[k] = fx(v);
    else {
      o[k] = fx(v);
      const dl = githubTreeDownload(o[k]);
      if (dl) o[`${k}_download`] = dl;
    }
  }
  return o;
};

export function find_talks() {
  for (const p of globSync("**/*.talk.y*ml", { cwd: here, absolute: true })) {
    const c = yaml.load(fs.readFileSync(p, "utf8"), { schema: yaml.JSON_SCHEMA /* keep dates unparsed */ });
    if (Array.isArray(c)) {
      talks.push(...c.map(t => fixlinks(t, p)));
    } else if (c && typeof c === "object") {
      for (const [n, t] of Object.entries(c)) {
        t.key = n;
        talks.push(fixlinks(t, p));
      }
    }
  }
  return talks
}

export function dump_talks(target_file) {
    return fs.writeFileSync(target_file, yaml.dump({ talks: find_talks() }))
}

// Usage like:
// console.log(find_talks())
